Multi folder push

nl0 opened the pull request:
# TODO

<!-- Use <del></del> for items that are not required for this PR -->

- [ ] Unit tests
- [ ] Automated tests (e.g. Preflight)
- [ ] Documentation
    - [ ] [Python: Run `build.py`](../gendocs/build.py) for new docstrings
    - [ ] JavaScript: basic explanation and screenshot of new features
- [x] [Changelog](CHANGELOG.md) entry

fiskus commented on catalog/app/containers/Bucket/PackageDialog/FilesInput.tsx:
```diff
@@ -375,78 +359,42 @@ const useFileStyles = M.makeStyles((t) => ({
-    [type, handle],
-  )
+type FileProps = FileOwnProps &
+  Omit<React.HTMLAttributes<HTMLDivElement>, keyof FileOwnProps>
```
even this

```tsx
interface FileProps extends React.HTMLAttributes<HTMLDivElement> {
  name: string
  state?: FilesEntryState
  size?: number
  action?: React.ReactNode
}
```

nl0 replied:
the point of this is to exclude props that are not passed to the div from the type to avoid possible conflicts and to articulate that those props arent related to the underlying html element

fiskus replied:
I wrote the same: "use HTMLAttributes from DivElement but re-define name, state size and action", didn't?

nl0 replied:
i was about to say that you cant override a prop of an interface with incompatible type, but seems like i was confusing declaration merging with plain interface extension, so you're probably right

fiskus commented on catalog/app/containers/Bucket/PackageDialog/FilesInput.tsx:
```diff
@@ -536,257 +477,69 @@ const useDirStyles = M.makeStyles((t) => ({
+    name,
+    state = 'unchanged',
+    active = false,
+    empty = false,
```
As I can see exposing `empty` is redundant because it uses `children.length`

nl0 replied:
there's a little nuance here: you may have empty children and not want to display the "<EMPTY>" placeholder  -- and that's exactly the case with the peer folders selection -- we dont support showing nested objects, so we just need to render the dir name and not its empty contents

fiskus commented on catalog/app/containers/Bucket/PackageDialog/FilesInput.tsx:
```diff
@@ -845,15 +591,449 @@ function DropzoneMessage({ error, warn, disabled }: DropzoneMessageProps) {
+  const classes = useLockStyles()
+  return (
+    <div className={classes.root}>
+      {!!progress && (
```
Early exit will make code slightly clearer:

```tsx
  const classes = useLockStyles()

  if (!progress) return <div className={classes.root} />

  return (
    <div className={classes.root}>
      <div className={classes.progressContainer}>
        …
      </div>
    </div>
  )
```

nl0 replied:
i dont think so bc in this particular case it adds repetition, while the condition just hides some nested elements, which seems idiomatic to me

fiskus commented on catalog/app/containers/Bucket/PackageDialog/FilesInput.tsx:
```diff
@@ -845,15 +591,449 @@ function DropzoneMessage({ error, warn, disabled }: DropzoneMessageProps) {
+
+const useFilesContainerStyles = M.makeStyles((t) => ({
+  root: {
+    direction: 'rtl', // show the scrollbar on the left
```
:(

nl0 replied:
?

fiskus replied:
It looks bad in my opinion. I hoped you would realize that after a while, but you stuck with this decision and scale it to another Dialog

nl0 replied:
the main reason for that is scrollbar on macos blocking the buttons on the right from being clicked, it's not an aesthetics-based decision, but more like a workaround for a usability issue, and i'm open to alternatives if you can provide any.

it looks okayish imo since most of the time the scrollbar is hidden (on macos at least).

fiskus replied:
The alternative is obvious, so I think you already declined it: show scrollbar at the right side, probably add some space with padding.

nl0 replied:
padding on the right looks as bad as the scrollbar on the left)
one alternative that comes to mind is (ab)using horizontal scrollbar (akin to smth like telegram embedded reader), but this seems even more odd and unconventional for a primarily desktop-oriented (web)app

fiskus commented on catalog/app/containers/Bucket/PackageDialog/FilesInput.tsx:
```diff
@@ -845,15 +591,449 @@ function DropzoneMessage({ error, warn, disabled }: DropzoneMessageProps) {
+  )
+
+  const onClick = React.useCallback((e: React.MouseEvent) => {
+    e.stopPropagation()
```
It worth writing down the reason in comments

nl0 replied:
`// stop click from propagating to parent elements and triggering their handlers`

fiskus replied:
The next team member will need to add some components above in the hierarchy, and he will have bad times finding why his `onClick` handler doesn't work. Finally, found this `stopPropagation` he will need some glue to refactor this and not to break already existing code. You can help him by providing the comment, explaining the reason behind this architectural decision.

If you don't understand why it's important, don't bother, we can get back to this when the problem becomes real if it does.

nl0 replied:
but i'm not arguing here -- i've added the comments)
or do you mean these comments should be more detailed?

fiskus replied:
I'm asking for
```js
// use Ohm's law to calculate voltage,
// see ohm-i.tsx and ohm-r.tsx
const u = i* r
```
rather then
```js
// assign `u` to the `i` multiplied by `r`, so `u` will be the product of `i` and `r` multiplication
const u = i * r
```
---
Again, if you don't understand what I mean, we can get back to this when the real problem occurs

fiskus commented on catalog/app/containers/Bucket/PackageDialog/FilesInput.tsx:
```diff
@@ -845,15 +591,449 @@ function DropzoneMessage({ error, warn, disabled }: DropzoneMessageProps) {
+  )
+
+  const onClick = React.useCallback((e: React.MouseEvent) => {
+    e.stopPropagation()
```
It blocks opening OS's file manager, I suppose? I think you should mention it in comment

fiskus replied:
There is `noClick` option for "react-dropzone". Probably, it will help

nl0 replied:
> It blocks opening OS's file manager, I suppose?

nope, it prevents the click from propagating to parent elements and triggering click handlers there.
adding the comment

fiskus commented on catalog/app/containers/Bucket/PackageDialog/FilesInput.tsx:
```diff
@@ -1016,88 +1195,207 @@ export function FilesInput({
+      </Header>
+
+      <ContentsContainer>
+        <Contents error={!!error} warn={truncated}>
```
I think it's better to rename `error` prop to something else, like `hasError` or `invalid`

nl0 replied:
why? MUI uses similar convention and i don's see what's wrong with it

fiskus replied:
It's harder to read and maintain code if identically named variables have different content

nl0 replied:
well, they have similar meaning

fiskus reviewed (approved):
Small UI issues:
* Directory has `cursor: pointer`, file has `cursor: default`
* Hard to see which entry is unselected, probably we can dim unselected one

![Screenshot from 2021-03-24 09-37-45](https://user-images.githubusercontent.com/533229/112272688-90a5e280-8c8d-11eb-8c07-44a1df9c4f95.png)

nl0 commented:
@fiskus 

> Directory has cursor: pointer, file has cursor: default

fixed

> Hard to see which entry is unselected, probably we can dim unselected one

yep, good idea, i've implemented that

nl0 commented on catalog/app/containers/Bucket/Dir.tsx:
```diff
@@ -25,9 +25,34 @@ import Summary from './Summary'
 import { displayError } from './errors'
 import * as requests from './requests'
 
+interface RouteMap {
```
it's the part of the route map relevant to the components in this file

> It's the auxiliary type, not the actual map, isn't it?

well, yeah, kinda

nl0 commented on catalog/app/containers/Bucket/PackageDialog/FilesInput.tsx:
```diff
@@ -259,6 +245,11 @@ export const validateHashingComplete = (state: FilesState) => {
 }
 
+export const validateNonEmptySelection = (state: FilesSelectorState) => {
+  if (state.every(R.propEq('selected', false))) return 'emptySelection'
```
ok
